File: python/xor_decrypt.py
import base64
from io import BytesIO
import hmac

from embit import bech32
from embit import compact

def xor_decrypt(key, payload):
    blob = base64.urlsafe_b64decode(payload)
    s = BytesIO(blob)
    variant = s.read(1)[0]
    if variant != 1:
        raise RuntimeError("Not implemented")
    # reading nonce
    l = s.read(1)[0]
    nonce = s.read(l)
    if len(nonce) != l:
        raise RuntimeError("Missing nonce bytes")
    if l < 8:
        raise RuntimeError("Nonce is too short")
    # reading payload
    l = s.read(1)[0]
    payload = s.read(l)
    if len(payload) > 32:
        raise RuntimeError("Payload is too long for this encryption method")
    if len(payload) != l:
        raise RuntimeError("Missing payload bytes")
    hmacval = s.read()
    expected = hmac.new(
        key, b"Data:" + blob[: -len(hmacval)], digestmod="sha256"
    ).digest()
    if len(hmacval) < 8:
        raise RuntimeError("HMAC is too short")
    if hmacval != expected[: len(hmacval)]:
        raise RuntimeError("HMAC is invalid")
    secret = hmac.new(key, b"Round secret:" + nonce, digestmod="sha256").digest()
    payload = bytearray(payload)
    for i in range(len(payload)):
        payload[i] = payload[i] ^ secret[i]
        print("payload[i]" + str(payload[i]));
    s = BytesIO(payload)
    pin = compact.read_from(s)
    amount_in_cent = compact.read_from(s)
    return pin, amount_in_cent

decryptionKey = b"shared-secret-key"
samplePayload = "AQhnxmlzUf9K7AWVl1HO17mbLmpwKbgl"
result = xor_decrypt(decryptionKey, samplePayload)
pin = result[0]
amount_in_cent = result[1]

print("pin=" + str(pin))
print("amount_in_cent=" + str(amount_in_cent))

File: node/xor_decrypt.js
const assert = require('assert');
const crypto = require('crypto');

const xor_decrypt = function(key, payload) {
    assert.ok(Buffer.isBuffer(key), new Error('Invalid argument ("key"): Buffer expected'));
    assert.ok(Buffer.isBuffer(payload), new Error('Invalid argument ("payload"): Buffer expected'));
    let pos = 0;
    const variant = payload[pos++];
    if (variant !== 1) {
        throw new Error('Not implemented');
    }
    const nonceLength = payload[pos++];
    const nonce = payload.subarray(pos++, pos += (nonceLength - 1));
    if (nonce.length != nonceLength) {
        throw new Error('Missing nonce bytes');
    }
    if (nonce.length < 8) {
        throw new Error('Nonce is too short');
    }
    const dataLength = payload[pos++];
    const data = payload.subarray(pos++, pos += (dataLength - 1));
    if (data.length > 32) {
        throw new Error('NPayload is too long for this encryption method');
    }
    if (data.length != dataLength) {
        throw new Error('Missing payload bytes');
    }
    const hmac = payload.subarray(pos);
    if (hmac.length < 8) {
        throw new Error('HMAC is too short');
    }
    const expected = crypto.createHmac('sha256', key).update(
        Buffer.concat([
            Buffer.from('Data:', 'utf8'),
            payload.subarray(0, payload.length - hmac.length)
        ])
    ).digest();
    if (hmac.compare(expected.subarray(0, hmac.length))) {
        throw new Error('HMAC is invalid');
    }
    const secret = crypto.createHmac('sha256', key).update(
        Buffer.concat([
            Buffer.from('Round secret:', 'utf8'),
            nonce
        ])
    ).digest();
    let buffer = Buffer.alloc(data.length);
    for (let index = 0; index < data.length; index++) {
        buffer[index] = data[index] ^ secret[index];
    }
    let offset = 0;
    // https://github.com/diybitcoinhardware/embit/blob/a43ee04d8619cdb4bbb84dab60002fd9c987ee60/src/embit/compact.py#L29-L36
    const readFrom = function(buffer) {
        int = buffer[offset++];
        if (int >= 0xFD) {
            const bytesToRead = 2 ** (int - 0xFC);
            int = buffer.readUIntLE(1, bytesToRead);
            offset += bytesToRead;
        }
        return int;
    };
    const pin = readFrom(buffer);
    const amount = readFrom(buffer);
    return { pin, amount };
};

const decryptionKey = Buffer.from('shared-secret-key', 'utf8');
const samplePayload = Buffer.from('AQhnxmlzUf9K7AWVl1HO17mbLmpwKbgl', 'base64');
const { pin, amount } = xor_decrypt(decryptionKey, samplePayload);
// `amount` is an integer number of cents.
console.log({ pin, amount });
